File: sdk/javascript/examples/descriptors/nem_mosaic.js
import symbolSdk from '../../src/index.js';

const { nem } = symbolSdk;

export default () => {
	const sampleAddress = 'TALICEROONSJCPHC63F52V6FY3SDMSVAEUGHMB7C';
	const textEncoder = new TextEncoder();

	// HACK: until fixed
	const levy = new nem.MosaicLevy();
	levy.transferFeeType = nem.MosaicTransferFeeType.ABSOLUTE;
	levy.recipientAddress.bytes = (new nem.Address(sampleAddress)).bytes;
	levy.mosaicId = new nem.MosaicId();
	levy.mosaicId.namespaceId = new nem.NamespaceId();
	levy.mosaicId.namespaceId.name = textEncoder.encode('lieutenant');
	levy.mosaicId.name = textEncoder.encode('colonel');
	levy.fee = new nem.Amount(632_0000n);

	return [
		// without properties
		{
			type: 'mosaic_definition_transaction_v1',
			rentalFeeSink: 'TBMOSAICOD4F54EE5CDMR23CCBGOAM2XSJBR5OLC',
			rentalFee: 50000n * 1000000n,

			mosaicDefinition: {
				ownerPublicKey: '00112233445566778899AABBCCDDEEFF00112233445566778899AABBCCDDEEFF',
				id: { namespaceId: { name: textEncoder.encode('genes') }, name: textEncoder.encode('memes') },
				description: textEncoder.encode('Not really valuable mosaic'),
				properties: [
				],
				levy: {}
			}
		},

		// with properties
		{
			type: 'mosaic_definition_transaction_v1',
			rentalFeeSink: 'TBMOSAICOD4F54EE5CDMR23CCBGOAM2XSJBR5OLC',
			rentalFee: 50000n * 1000000n,

			mosaicDefinition: {
				ownerPublicKey: '00112233445566778899AABBCCDDEEFF00112233445566778899AABBCCDDEEFF',
				id: { namespaceId: { name: textEncoder.encode('genes') }, name: textEncoder.encode('memes') },
				description: textEncoder.encode('Not really valuable mosaic'),
				properties: [
					{ property: { name: textEncoder.encode('divisibility'), value: textEncoder.encode('3') } },
					{ property: { name: textEncoder.encode('initialSupply'), value: textEncoder.encode('123_000') } },
					{ property: { name: textEncoder.encode('supplyMutable'), value: textEncoder.encode('false') } },
					{ property: { name: textEncoder.encode('transferable'), value: textEncoder.encode('true') } }
				],
				levy: {}
			}
		},

		// with levy
		{
			type: 'mosaic_definition_transaction_v1',
			rentalFeeSink: 'TBMOSAICOD4F54EE5CDMR23CCBGOAM2XSJBR5OLC',
			rentalFee: 50000n * 1000000n,

			mosaicDefinition: {
				ownerPublicKey: '00112233445566778899AABBCCDDEEFF00112233445566778899AABBCCDDEEFF',
				id: { namespaceId: { name: textEncoder.encode('genes') }, name: textEncoder.encode('memes') },
				description: textEncoder.encode('Not really valuable mosaic'),
				properties: [
				],
				levy: {
					transferFeeType: 'absolute',
					recipientAddress: sampleAddress,
					mosaicId: { namespaceId: { name: textEncoder.encode('lieutenant') }, name: textEncoder.encode('colonel') },
					fee: 632_0000n
				}
			}
		},

		// supply change
		{
			type: 'mosaic_supply_change_transaction_v1',
			mosaicId: { namespaceId: { name: textEncoder.encode('genes') }, name: textEncoder.encode('memes') },
			action: 'increase',
			delta: 321_000n
		}
	];
};

File: sdk/python/examples/descriptors/nem_mosaic.py
from symbolchain.facade.NemFacade import Address
from symbolchain.nc import Address as sc_Address
from symbolchain.nc import Amount, MosaicId, MosaicLevy, MosaicTransferFeeType, NamespaceId


def descriptor_factory():
	sample_address = 'TALICEROONSJCPHC63F52V6FY3SDMSVAEUGHMB7C'

	# HACK: until fixed
	levy = MosaicLevy()
	levy.transfer_fee_type = MosaicTransferFeeType.ABSOLUTE
	levy.recipient_address = sc_Address(str(Address(sample_address)).encode('utf8'))
	levy.mosaic_id = MosaicId()
	levy.mosaic_id.namespace_id = NamespaceId()
	levy.mosaic_id.namespace_id.name = 'lieutenant'.encode('utf8')
	levy.mosaic_id.name = 'colonel'.encode('utf8')
	levy.fee = Amount(632_0000)

	return [
		# without properties
		{
			'type': 'mosaic_definition_transaction_v1',
			'rental_fee_sink': 'TBMOSAICOD4F54EE5CDMR23CCBGOAM2XSJBR5OLC',
			'rental_fee': 50000 * 1000000,

			'mosaic_definition': {
				'owner_public_key': '00112233445566778899AABBCCDDEEFF00112233445566778899AABBCCDDEEFF',
				'id': {'namespace_id': {'name': 'genes'.encode('utf8')}, 'name': 'memes'.encode('utf8')},
				'description': 'Not really valuable mosaic'.encode('utf8'),
				'properties': [
				],
				'levy': {}
			}
		},

		# with properties
		{
			'type': 'mosaic_definition_transaction_v1',
			'rental_fee_sink': 'TBMOSAICOD4F54EE5CDMR23CCBGOAM2XSJBR5OLC',
			'rental_fee': 50000 * 1000000,

			'mosaic_definition': {
				'owner_public_key': '00112233445566778899AABBCCDDEEFF00112233445566778899AABBCCDDEEFF',
				'id': {'namespace_id': {'name': 'genes'.encode('utf8')}, 'name': 'memes'.encode('utf8')},
				'description': 'Not really valuable mosaic'.encode('utf8'),
				'properties': [
					{'property_': {'name': 'divisibility'.encode('utf8'), 'value': '3'.encode('utf8')}},
					{'property_': {'name': 'initialSupply'.encode('utf8'), 'value': '123_000'.encode('utf8')}},
					{'property_': {'name': 'supplyMutable'.encode('utf8'), 'value': 'false'.encode('utf8')}},
					{'property_': {'name': 'transferable'.encode('utf8'), 'value': 'true'.encode('utf8')}}
				],
				'levy': {}
			}
		},

		# with levy
		{
			'type': 'mosaic_definition_transaction_v1',
			'rental_fee_sink': 'TBMOSAICOD4F54EE5CDMR23CCBGOAM2XSJBR5OLC',
			'rental_fee': 50000 * 1000000,

			'mosaic_definition': {
				'owner_public_key': '00112233445566778899AABBCCDDEEFF00112233445566778899AABBCCDDEEFF',
				'id': {'namespace_id': {'name': 'genes'.encode('utf8')}, 'name': 'memes'.encode('utf8')},
				'description': 'Not really valuable mosaic'.encode('utf8'),
				'properties': [
				],
				'levy': {
					'transfer_fee_type': 'absolute',
					'recipient_address': sample_address,
					'mosaic_id': {'namespace_id': {'name': 'lieutenant'.encode('utf8')}, 'name': 'colonel'.encode('utf8')},
					'fee': 632_0000
				}
			}
		},

		# supply change
		{
			'type': 'mosaic_supply_change_transaction_v1',
			'mosaic_id': {'namespace_id': {'name': 'genes'.encode('utf8')}, 'name': 'memes'.encode('utf8')},
			'action': 'increase',
			'delta': 321_000
		}
	]
